### PR TITLE
Speed up Map.Contains checks for non-flat maps.

### PR DESCRIPTION
**Pull request by RoosterDragon**
We observe that most cells within a map lie within a region where no matter their height, their projection would still remain in map bounds. We can utilise this to perform a fast check for such cells and skipping the expensive checks on their actual height. We only need to check the actual height of a cell if this could cause the projection to go out of bounds.

In this shoddy diagram, the map bounds are given in blue and the projection-safe region given in green. Any cell in the green region, no matter it's height, would still remain in bounds after projection. The red line is effectively the maximum height it can gain (which is Grid.MaximumTerrainHeight + 2 for reasons described in the code comments)

This means for anything in the green region, i.e. most of the map, we can do a much faster `Map.Contains` check on non-flat maps. We only need to fall back to the slow check for stuff on the extreme left/right edge or in the buffer near the top of the map, where the actual height of the cells means it's projection may or may not be within the map bounds.

image

### CODE REVIEW

**Review by reaperrr — APPROVED** [review]
Code makes sense to me and didn't notice any issues in some TS testing

File: OpenRA.Game/Map/Map.cs [hunk 1]
```diff
@@ -838,6 +840,14 @@ bool ContainsAllProjectedCellsCovering(MPos uv)
 			if (Grid.MaximumTerrainHeight == 0)
 				return Bounds.Contains(uv.U, uv.V);
 
+			// PERF: Most cells lie within a region where no matter their height,
+			// all possible projected cells would remain in the map area.
+			// For these, we can do a fast-path check.
+			if (projectionSafeBounds.Contains(uv.U, uv.V))
```
Comment (pchote) [inline OpenRA.Game/Map/Map.cs:846]: It looks like this supersedes the `Grid.MaximumTerrainHeight == 0` check above without any additional performance overheads, so we can now remove that?

File: OpenRA.Game/Map/Map.cs [hunk 2]
```diff
@@ -1029,6 +1039,22 @@ public void SetBounds(PPos tl, PPos br)
...
+			// V: For an even height tile, this ranges from 0 to height
+			//    For an odd tile, the height may get rounded up to next even.
+			//    Then also it projects to four tiles which adds one more to the possible height change.
+			//    So we get a range of 0 to height + 1 + 1.
+			//    As the height only goes upwards, we only need to make room at the top of the map and not the bottom.
+			var maxHeight = Grid.MaximumTerrainHeight;
+			if ((maxHeight & 1) == 1)
+				maxHeight += 2;
+			projectionSafeBounds = Rectangle.FromLTRB(
+				Bounds.Left + 1,
```
Comment (pchote) [inline OpenRA.Game/Map/Map.cs:1053]: We only need to inset these if the height is odd, right?

File: OpenRA.Game/Map/Map.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Game/Map/Map.cs:846]: Ah, actually not for the left- and right-most columns of the map.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c36de75e1ecb944165a34a005bcbbf96e0d7e8ff

**Pull request merged**